### DISCUSSION TITLE
Improvements for opening files

### DISCUSSION THREAD

**Discussion by felixgabler** [flyerhq/flutter_chat_ui#242] [Ideas & Requests]
Hi! Hope you are safe and doing well.

I just implemented file sharing and opening as described in the [docs](https://docs.flyer.chat/flutter/chat-ui/advanced-usage#opening-a-file). For the most part, I am very happy with it but I do think it can be improved in two regards:

1. [DOCS] Add information about opening remote URLs to the docs. One also needs to download files e.g. using `flutter_cache_manager` or something similar. The code then would be:
```
final file = await DefaultCacheManager().getSingleFile(url);
OpenFile.open(file.path);
``` 

2. [FEATURE] Because downloading and showing the file can take some time, it would be awesome if we could show a loading spinner inside of the message. I am imagining a circle spinner inside the beautiful file icon area you have designed. Do let me know if I should make a PR for this to showcase what I am thinking of.

Best wishes again!

**Comment by demchenkoalex** [comment]
Hi, all good thank you!

Regarding docs - they are only an example on what can you do with a chat. It is dependency agnostic when it comes to files, I don't want to write docs to all the libraries out there :) So devs will need to check out the documentation of the files library they are using.

Regarding number 2 - you could totally do it, we will benefit from it :) you should have everything needed to do it, you can use `CircularProgressIndicator` instead of the file icon with the colour of the file icon when downloading a file. I guess we will need to add a parameter to the file message itself?

**Reply by felixgabler** [reply]
Perhaps we could add a hint in the doc's code example though about the fact that `message.uri` would need to be a local path in order for it to open. I believe the use case for most people is saving messages (and files) somewhere in the cloud or on servers. If you propose `open_files` one should at least also hint at its inability to open remote files without downloading them. I had to look at old issues in the `open_files` library to figure this out myself.

**Reply by demchenkoalex** [reply]
You should have looked at the example in Firebase repo :) https://github.com/flyerhq/flutter_firebase_chat_core/blob/main/example/lib/chat.dart#L145

**Reply by Faaatman** [reply]
Do you guys think we can implement a file downloading system using [Flutter Downloader](https://pub.dev/packages/flutter_downloader)?
This would help us implement the following on both FileMessage and ImageMessage:

- loading indicator which we don't have to set ourselves (or download progress indicator?)
- a status button (download or retry if failed)

It already uses [OpenFile](https://pub.dev/packages/open_file) to open files.

Upon adding the message we can add a new task for it in flutter downloader and download it immediately?

**Reply by felixgabler** [reply]
Thanks for the suggestion! I think that dependency is quite heavy and I am not sure how many people even use file messages ATM. I doubt that @demchenkoalex will want to add it into the main plugin. Perhaps it would be cool to add it to the example though and refer to it in the docs.

However, it could also become interesting with regards to the discussion we had about multiple images and downloading for that. A clean system for this logic could lift this plugin up a lot.

**Reply by demchenkoalex** [reply]
Definitely will not be adding this to the package itself, now it is somewhat working on Android, iOS, Linux, macOS, Web and Windows and adding this will result in locking to Android and iOS. Even though I know the chat will not work 100% on not mobile platforms, it is nice to have still, at least for plain text messages.

**Reply by Faaatman** [reply]
Thank you for your answers
